### PR TITLE
Fix solana dapp swap

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/23594

This PR resolves the compatibility issues where dApp uses [Transaction.partialSign](https://github.com/solana-labs/solana-web3.js/blob/5790271d25f61edc535c7bf187f8d5754e79e2b8/src/transaction.ts#L628) or [Transaction.AddSignature](https://github.com/solana-labs/solana-web3.js/blob/5790271d25f61edc535c7bf187f8d5754e79e2b8/src/transaction.ts#L666) to provide signatures for other signer accounts needed in transactions.
With this PR, we get Transaction.signatures and use those provided signatures when serializing transactions. For fee payer signature, sign the encoded serialized message from dApp directly, because we need to sign the exact serialized message as they use for other provided signatures.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Try swap from SOL to USDC on both https://raydium.io/swap/ and https://www.orca.so/ on Solana mainnet, both should work fine.

### CODE REVIEW

File: components/brave_wallet/browser/solana_transaction.h [hunk 1]
```diff
@@ -68,7 +70,12 @@ class SolanaTransaction {
   // Serialize and encode the message in Base64.
   std::string GetBase64EncodedMessage() const;
   absl::optional<std::vector<uint8_t>> GetSignedTransactionBytes(
-      KeyringService* keyring_service) const;
+      KeyringService* keyring_service,
+      std::vector<uint8_t>* fee_payer_signature = nullptr) const;
```
Comment (yrliou) [inline components/brave_wallet/browser/solana_transaction.h:75]: fee_payer_signature is added for hardware transaction to passed in signature.

File: components/brave_wallet/browser/solana_transaction.cc [hunk 2]
```diff
@@ -176,35 +182,78 @@ bool SolanaTransaction::operator!=(const SolanaTransaction& tx) const {
...
-  if (!message_bytes || signers.empty())
+  auto message_signers_pair = GetSerializedMessage();
+  if (!message_signers_pair)
     return absl::nullopt;
+  auto& message_bytes = message_signers_pair->first;
+  auto& signers = message_signers_pair->second;
 
   std::vector<uint8_t> transaction_bytes;
   // Compact array of signatures.
   CompactU16Encode(signers.size(), &transaction_bytes);
```
Comment (darkdh) [inline components/brave_wallet/browser/solana_transaction.cc:229]: This should be exactly same size as the total signatures. If total signatures is less than this size, when a dApp tries to use byte array to construct a web3 Transaction object, that operation might fail.

File: components/brave_wallet/browser/solana_transaction_unittest.cc [hunk 3]
```diff
@@ -515,4 +508,35 @@ TEST_F(SolanaTransactionUnitTest, GetBase64EncodedMessage) {
...
+                              {system_transfer_ins()});
+  auto expected_message_bytes = tx.message()->Serialize(nullptr);
+  ASSERT_TRUE(expected_message_bytes);
+  EXPECT_EQ(
+      tx.GetSerializedMessage(),
+      std::make_pair(*expected_message_bytes,
+                     std::vector<std::string>(
+                         {"BrG44HdsEhzapvs8bEqzvkq4egwevS3fRE6ze2ENo6S8"})));
+
+  // Create test message bytes by modifying instruction data part.
```
Comment (darkdh) [inline components/brave_wallet/browser/solana_transaction_unittest.cc:547]: Can we add a comment saying this is to create a valid message bytes which is different than `expected_message_bytes`?

File: browser/brave_wallet/solana_provider_browsertest.cc [hunk 4]
```diff
@@ -256,6 +312,17 @@ class SolanaProviderTest : public InProcessBrowserTest {
     run_loop.Run();
   }
 
+  void AddAccount2() {
```
Comment (darkdh) [inline browser/brave_wallet/solana_provider_browsertest.cc:315]: This is unused.

File: components/brave_wallet/browser/solana_transaction.cc (hunk 2 above)
Comment (darkdh) [reply components/brave_wallet/browser/solana_transaction.cc:229]: Also in https://docs.solana.com/developing/programming-model/transactions#transaction-format
`t also verifies that each signature was signed by the private key corresponding to the public key at the same index in the message's account addresses array`
So we should follow the order of signer in message.

File: browser/brave_wallet/solana_provider_browsertest.cc (hunk 4 above)
Comment (yrliou) [reply browser/brave_wallet/solana_provider_browsertest.cc]: Removed in https://github.com/brave/brave-core/pull/14001/commits/4ee63086cdc0d66ea8b7845fceb6eecead0567ce

File: components/brave_wallet/browser/solana_transaction_unittest.cc (hunk 3 above)
Comment (yrliou) [reply components/brave_wallet/browser/solana_transaction_unittest.cc]: This is gone now in https://github.com/brave/brave-core/pull/14001/commits/4ee63086cdc0d66ea8b7845fceb6eecead0567ce.

File: components/brave_wallet/common/brave_wallet.mojom [hunk 5]
```diff
@@ -142,6 +142,16 @@ interface EthereumProvider {
   IsLocked() => (bool isLocked);
 };
 
+struct SignaturePubkeyPair {
```
[the remark this answers is not in the document]
Comment (yrliou) [reply components/brave_wallet/common/brave_wallet.mojom:146]: Added in https://github.com/brave/brave-core/pull/14001/commits/4ee63086cdc0d66ea8b7845fceb6eecead0567ce

File: components/brave_wallet/browser/solana_transaction.cc (hunk 2 above)
Comment (yrliou) [reply components/brave_wallet/browser/solana_transaction.cc]: Use actual size of total signatures: https://github.com/brave/brave-core/pull/14001/commits/30f3b6d3cc5e2822bdea392e9d4deb1cad59ecb4
Fix order of signer: https://github.com/brave/brave-core/pull/14001/commits/4ee63086cdc0d66ea8b7845fceb6eecead0567ce

File: components/brave_wallet/browser/solana_transaction_unittest.cc [hunk 6]
```diff
@@ -209,6 +209,18 @@ TEST_F(SolanaTransactionUnitTest, GetSignedTransaction) {
...
+  sign_tx_param->signatures.pop_back();
+  transaction2.set_sign_tx_param(sign_tx_param.Clone());
+  expected_bytes = std::vector<uint8_t>({1});  // # of signature
+  expected_bytes.insert(expected_bytes.end(), signature.begin(),
+                        signature.end());
+  expected_bytes.insert(expected_bytes.end(), message_bytes.begin(),
+                        message_bytes.end());
+  EXPECT_EQ(transaction2.GetSignedTransaction(keyring_service()),
+            base::Base64Encode(expected_bytes));
+
```
Comment (darkdh) [inline components/brave_wallet/browser/solana_transaction_unittest.cc:231]: can we also add a test that is num of signatures larger than signers.size? And final signature size should be signers.size.

Comment (yrliou) [reply components/brave_wallet/browser/solana_transaction_unittest.cc:240]: https://github.com/brave/brave-core/compare/4ee63086cdc0d66ea8b7845fceb6eecead0567ce..42a28ae80d207a4b3b9cb1e82133028e8f30559d

**Review by darkdh — APPROVED** [review]

**Pull request merged**